### PR TITLE
[Inductor][CPU] Add GEMM templates for _weight_int4pack_mm_for_cpu with AMX

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150603

**Summary**
It's part of the task to enable max-autotune with GEMM template for WoQ INT4 GEMM on CPU.

This PR adds AMX-based GEMM templates for `torch.ops.aten_weight_int4pack_mm_for_cpu`. It brings performance benefits on platforms where AMX is available.

**Validation results**
We have run GPT-J-6B and Llama-3-8B-Instruct on a 6th gen Xeon with 96 cores. Results show that the AMX-based microkernel outperforms AVX512-based one by >5x for prefill stage with 1024 input length.

**Test plan**
```
python test/inductor/test_cpu_select_algorithm.py -k test_int4_woq_mm_amx
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Comment by Xia-Weiwen** [comment]
Hello @laithsakka I met a problem with `pr_time_benchmarks` and may need your help.
This PR is having a CI failure which says:
```
REGRESSION: benchmark ('aotdispatcher_training_subclass_cpu', 'compile_time_instruction_count') failed,
actual result 10040358265 is 2.14% higher than expected 9830000000 ±+1.50% if this is an expected regression,
please update the expected results.
```
I have tried rerunning the benchmark but all failed with different actual results.

Then I found the notes here: https://github.com/pytorch/pytorch/blob/main/benchmarks/dynamo/pr_time_benchmarks/README.md

I followed the notes to run the failed benchmark with main branch (commit=164d2c887b45ec2673b9bbc08e457bfab0a1eb3c on Apr. 5) and got the following results:
```
$ cd benchmarks/dynamo/pr_time_benchmarks/
$ PYTHONPATH=./ python benchmarks/aotdispatcher.py a.txt
collecting compile time instruction count for aotdispatcher_training_subclass_cpu
compile time instruction count for iteration 0 is 10346174811
compile time instruction count for iteration 1 is 10344511496
compile time instruction count for iteration 2 is 10344707896
compile time instruction count for iteration 3 is 10342844652
compile time instruction count for iteration 4 is 10342457766
```
The results are larger than expected as well. So, it looks like the regression is not caused by this PR. However, it's weird that other PRs do not have such CI failures.
The notes say I should contact you about it. Do you think the expected results of this benchmark should be updated? Or is there another way to resolve the failure? Thanks!

**Review by leslie-fang-intel — COMMENTED** [review]
Please also provide the perf data for this optimization.

File: test/inductor/test_cpu_select_algorithm.py [hunk 1]
```diff
@@ -1524,7 +1524,7 @@ def forward(self, a):
     @patches
     @torch.no_grad
     @dtypes(torch.bfloat16)
-    @parametrize("batch_size", (32,))
+    @parametrize("batch_size", (1,))
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply test/inductor/test_cpu_select_algorithm.py:1527]: Because otherwise it will go to the AMX GEMM template.

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 2]
```diff
@@ -1139,7 +1143,11 @@ def pack_vnni_weight(cls, W, micro_gemm, new_size):
         else:
             k = new_size[-2]
             # Apply VNNI packing to the weight tensor
-            if micro_gemm.get_b_layout() != LayoutType.NORMAL:
+            # WOQ INT4 weights are reordered in microkernel
+            if (
+                micro_gemm.get_b_layout() != LayoutType.NORMAL
+                and not micro_gemm.is_woq_int4()
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply torch/_inductor/codegen/cpp_gemm_template.py:1149]: May I know what the helper function should do?

[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply torch/_inductor/codegen/cpp_gemm_template.py:1149]: Since these checks are invoked twice with the same conditions, we can extract them into a helper function

File: torch/_inductor/codegen/cpp_micro_gemm.py [hunk 3]
```diff
@@ -1179,11 +1179,15 @@ def codegen_define(self, kernel: CppTemplateKernel) -> str:
             assert block_k == 64, "Only support block_k = 64 for AMX INT8"
         else:
             assert block_k == 32, "Only support block_k = 32 for AMX Bfloat16/Float16"
-        num_columns = block_n // 16
+        # block_n for woq int4 is 64, which is too large for micro kernel
+        # so we split it into 2x32
+        num_columns = 2 if self.is_woq_int4() else block_n // 16
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply torch/_inductor/codegen/cpp_micro_gemm.py]: Thanks. Changed.

File: torch/_inductor/codegen/cpp_micro_gemm.py [hunk 4]
```diff
@@ -1572,6 +1576,267 @@ def is_woq_int4(self):
...
+    {{kernel.assert_function}}(K % 2 == 0, "K dimension must be multiple of 2");
+    {{kernel.assert_function}}({{block_n}} == 64, "block_n must be 64 for WOQ int4");
+
+    // Create a stack-allocated buffer for tiles of B.
+    // Except maybe for the tail-case, an AMX tile of B has 16x32 BF16 elements.
+    // we cache K * {{block_n}} elements of dequantized B
+    {{template.codegen_allocate_weight_buffer("dequantized_B_buf", input_t, "K", block_n)}}
+
+    constexpr int BLOCK_K = {{block_k}};
+    constexpr int64_t BLOCK_N = 64;
```
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply torch/_inductor/codegen/cpp_micro_gemm.py]: Thanks. Changed.

File: torch/_inductor/codegen/cpp_gemm_template.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply torch/_inductor/codegen/cpp_gemm_template.py]: Sure. I have simplified this.

**Comment by Xia-Weiwen** [comment]
> Please also provide the perf data for this optimization.

Added. Please review again. Thanks.

**Review by leslie-fang-intel — APPROVED** [review]

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/cpp_gemm_template.py [hunk 5]
```diff
@@ -1115,14 +1115,20 @@ def block_weight(cls, W, new_size, padding):
 
     @classmethod
     def pack_vnni_weight(cls, W, micro_gemm, new_size):
+        # WOQ INT4 weights are reordered in microkernel so do not pack them here
+        should_pack = (
+            micro_gemm.get_b_layout() != LayoutType.NORMAL
+            and not micro_gemm.is_woq_int4()
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_gemm_template.py:1121]: If the packing happens at the runtime (in the micro-kernel), should we set the layout of "b" to be NORMAL instead?

Comment (Xia-Weiwen) [reply torch/_inductor/codegen/cpp_gemm_template.py:1121]: Thanks for the comments. Layout of b is not normal. It's blocked but not VNNI. At runtime, we load two lines of K dim and shuffle them to VNNI layout for computation.
